File: React/src/services/aiService.ts
import type { ApiResult } from '@models/api';
import type { ScheduleOptimizationDto, AppointmentAutoFillDto } from '@models/dtos';

export interface AIRequest {
  workflowId: string;
  context: Record<string, unknown>;
}

export interface AIResponse<T> {
  suggestions: T;
  confidence: number;
  explanation: string;
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

export async function scheduleOptimization(
  context: { providerId?: string; date?: string; departmentId?: string }
): Promise<ApiResult<ScheduleOptimizationDto>> {
  await sleep(600 + Math.floor(Math.random() * 400));

  const suggestions: ScheduleOptimizationDto['suggestions'] = [
    {
      type: 'fill',
      appointmentId: null,
      proposedStart: `${context.date ?? '2026-06-15'}T09:00:00Z`,
      proposedEnd: `${context.date ?? '2026-06-15'}T09:30:00Z`,
      providerId: context.providerId ?? 'provider-1',
      providerName: 'Dr. Sarah Chen',
      reason: 'Fill 30-min gap with waitlist patient',
      estimatedUtilizationGain: 0.12,
      estimatedNoShowReduction: 0.03,
      waitlistMatches: [
        { waitlistId: 'wl-101', patientName: 'James Wilson', fitScore: 0.92 },
        { waitlistId: 'wl-102', patientName: 'Maria Garcia', fitScore: 0.85 },
      ],
    },
    {
      type: 'fill',
      appointmentId: null,
      proposedStart: `${context.date ?? '2026-06-15'}T14:00:00Z`,
      proposedEnd: `${context.date ?? '2026-06-15'}T14:15:00Z`,
      providerId: context.providerId ?? 'provider-2',
      providerName: 'Dr. Michael Torres',
      reason: 'Fill 15-min gap between existing appointments',
      estimatedUtilizationGain: 0.08,
      estimatedNoShowReduction: 0.02,
      waitlistMatches: [
        { waitlistId: 'wl-103', patientName: 'Robert Kim', fitScore: 0.78 },
      ],
    },
  ];

  const result: ScheduleOptimizationDto = {
    suggestions,
    confidence: 0.87,
    explanation: `Found ${suggestions.length} schedule gaps on ${context.date ?? '2026-06-15'}. Filling gaps with waitlist patients could improve utilization by ${(suggestions.reduce((s, i) => s + i.estimatedUtilizationGain, 0) * 100).toFixed(0)}%.`,
  };

  return { status: 'ok', data: result };
}

export async function appointmentSuggest(
  context: { patientId: string; preferredDepartmentId?: string; reasonHint?: string }
): Promise<ApiResult<AppointmentAutoFillDto>> {
  await sleep(500 + Math.floor(Math.random() * 300));

  const result: AppointmentAutoFillDto = {
    suggestions: {
      appointmentType: 'Follow-Up',
      providerId: 'provider-1',
      proposedDateTime: '2026-06-22T10:00:00Z',
      durationMinutes: 15,
      reasonForVisit: context.reasonHint ?? 'Follow-up lab results',
    },
    confidence: 0.88,
    explanation: 'Patient typically books 15-min follow-ups with Dr. Chen every 4 weeks.',
  };

  return { status: 'ok', data: result };
}

File: WebAPI/HealthcareAppointmentOps.Application/Services/AiService.cs
using HealthcareAppointmentOps.Application.Abstractions;
using HealthcareAppointmentOps.Application.DTOs;
using HealthcareAppointmentOps.Application.Responses;
using HealthcareAppointmentOps.Domain.Entities;

namespace HealthcareAppointmentOps.Application.Services;

public class AiService : IAiService
{
    private readonly IAppointmentRepository _appointmentRepo;
    private readonly IWaitlistRepository _waitlistRepo;
    private readonly IProviderRepository _providerRepo;

    public AiService(
        IAppointmentRepository appointmentRepo,
        IWaitlistRepository waitlistRepo,
        IProviderRepository providerRepo)
    {
        _appointmentRepo = appointmentRepo;
        _waitlistRepo = waitlistRepo;
        _providerRepo = providerRepo;
    }

    public async Task<ApiResult<ScheduleOptimizationDto>> OptimizeScheduleAsync(ScheduleOptimizationRequest request, CancellationToken ct = default)
    {
        var date = request.Date ?? DateOnly.FromDateTime(DateTime.UtcNow.AddDays(1));
        var dateTime = date.ToDateTime(TimeOnly.MinValue);
        var endDateTime = dateTime.AddDays(1);

        var appointments = await _appointmentRepo.GetByDateRangeAsync(dateTime, endDateTime, ct);
        if (request.ProviderId.HasValue)
            appointments = appointments.Where(a => a.ProviderId == request.ProviderId.Value).ToList();
        if (request.DepartmentId.HasValue)
            appointments = appointments.Where(a => a.DepartmentId == request.DepartmentId.Value).ToList();

        var providers = await _providerRepo.GetAllAsync(ct);
        if (request.ProviderId.HasValue)
            providers = providers.Where(p => p.ProviderId == request.ProviderId.Value).ToList();
        if (request.DepartmentId.HasValue)
            providers = providers.Where(p => p.DepartmentId == request.DepartmentId.Value).ToList();

        var (waitlistItems, _) = await _waitlistRepo.ListAsync("Open", request.DepartmentId, 0, 100, ct);

        var suggestions = new List<ScheduleOptimizationSuggestionDto>();
        var random = new Random(42); // deterministic seed

        foreach (var provider in providers)
        {
            var providerAppts = appointments.Where(a => a.ProviderId == provider.ProviderId && a.Status != "Cancelled" && a.Status != "NoShow").OrderBy(a => a.ScheduledDateTime).ToList();
            if (providerAppts.Count < 2) continue;

            for (int i = 0; i < providerAppts.Count - 1; i++)
            {
                var currentEnd = providerAppts[i].ScheduledDateTime.AddMinutes(providerAppts[i].DurationMinutes);
                var nextStart = providerAppts[i + 1].ScheduledDateTime;
                var gapMinutes = (nextStart - currentEnd).TotalMinutes;

                if (gapMinutes >= 15)
                {
                    var matches = waitlistItems.Take(2).Select((w, idx) => new ScheduleOptimizationMatchDto
                    {
                        WaitlistId = w.WaitlistId.ToString(),
                        PatientName = w.Patient != null ? $"{w.Patient.FirstName} {w.Patient.LastName}" : "Patient",
                        FitScore = Math.Round(0.92 - (idx * 0.07), 2)
                    }).ToList();

                    if (matches.Count > 0)
                    {
                        suggestions.Add(new ScheduleOptimizationSuggestionDto
                        {
                            Type = "fill",
                            AppointmentId = null,
                            ProposedStart = currentEnd,
                            ProposedEnd = nextStart,
                            ProviderId = provider.ProviderId,
                            ProviderName = $"{provider.FirstName} {provider.LastName}",
                            Reason = $"Fill {gapMinutes:F0}-min gap with waitlist patient",
                            EstimatedUtilizationGain = Math.Round(random.NextDouble() * 0.15, 2),
                            EstimatedNoShowReduction = Math.Round(random.NextDouble() * 0.05, 3),
                            WaitlistMatches = matches
                        });
                    }
                }
            }
        }

        var confidence = Math.Min(0.95, 0.7 + suggestions.Count * 0.05);
        var explanation = suggestions.Count > 0
            ? $"Found {suggestions.Count} schedule gap(s) on {date:yyyy-MM-dd}. Filling gaps with waitlist patients could improve utilization."
            : $"No schedule gaps found on {date:yyyy-MM-dd}. Schedule is well-optimized.";

        return ApiResult<ScheduleOptimizationDto>.Success(new ScheduleOptimizationDto
        {
            Suggestions = suggestions.Take(5).ToList(),
            Confidence = Math.Round(confidence, 2),
            Explanation = explanation
        });
    }

    public async Task<ApiResult<AppointmentAutoFillDto>> SuggestAppointmentAsync(AppointmentSuggestRequest request, CancellationToken ct = default)
    {
        var patientAppts = await _appointmentRepo.GetByPatientIdAsync(request.PatientId, ct);
        var lastAppt = patientAppts.FirstOrDefault();

        string appointmentType = "Follow-Up";
        Guid providerId = Guid.Empty;
        int durationMinutes = 15;
        string reason = request.ReasonHint ?? "Follow-up visit";

        if (lastAppt != null)
        {
            appointmentType = lastAppt.AppointmentType;
            providerId = lastAppt.ProviderId;
            durationMinutes = lastAppt.DurationMinutes;
        }

        if (providerId == Guid.Empty)
        {
            var providers = await _providerRepo.GetAllAsync(ct);
            providerId = providers.FirstOrDefault()?.ProviderId ?? Guid.Empty;
        }

        var proposedDateTime = DateTime.UtcNow.AddDays(14).Date.AddHours(10);

        var confidence = lastAppt != null ? 0.88 : 0.72;
        var explanation = lastAppt != null
            ? "Patient typically books follow-ups with their last provider every 2-4 weeks."
            : "Suggested based on common follow-up patterns and provider availability.";

        return ApiResult<AppointmentAutoFillDto>.Success(new AppointmentAutoFillDto
        {
            Suggestions = new AppointmentAutoFillSuggestionDto
            {
                AppointmentType = appointmentType,
                ProviderId = providerId,
                ProposedDateTime = proposedDateTime,
                DurationMinutes = durationMinutes,
                ReasonForVisit = reason
            },
            Confidence = Math.Round(confidence, 2),
            Explanation = explanation
        });
    }
}
